### PR TITLE
fix: concurrency problems at local metadata store

### PR DESCRIPTION
**Pull request by fraidev**
This PR must solve the weird deadlocks happening after resuming local cluster with previous data.

Solves: https://github.com/infinyon/fluvio/issues/4162

### CODE REVIEW

File: crates/fluvio-stream-dispatcher/src/metadata/local.rs [hunk 1]
```diff
@@ -309,15 +308,17 @@ cfg_if::cfg_if! {
...
-                let read = self.stores.upgradable_read().await;
+                let read = self.stores.read();
                 Ok(match read.get(key) {
                     Some(store) => store.clone(),
                     None => {
-                        let mut write = RwLockUpgradableReadGuard::upgrade(read).await;
-                        let store = Arc::new(SpecStore::load::<S, _>(self.path.join(key)).await?);
+                        drop(read);
+                        let mut write = self.stores.write();
+                        let store = Arc::new(SpecStore::load::<S, _>(self.path.join(key))?);
```
Comment (sehz) [inline crates/fluvio-stream-dispatcher/src/metadata/local.rs:319]: we need to move this out of here but it's not blocker.  let's create issue for it

Comment (fraidev) [reply crates/fluvio-stream-dispatcher/src/metadata/local.rs:319]: Good catch! I forgot to remove it

File: crates/fluvio-stream-dispatcher/src/metadata/local.rs [hunk 2]
```diff
@@ -585,7 +597,9 @@ cfg_if::cfg_if! {
 
             fn flush<S: Spec>(&self) -> Result<()> {
                 let storage: VersionedSpecStorage<S> = self.try_into()?;
-                serde_yaml::to_writer(std::fs::File::create(&self.path)?, &storage)?;
+                let file = std::fs::File::create(&self.path)?;
+                serde_yaml::to_writer(&file, &storage)?;
+                file.sync_all()?;
```
[the remark this answers is not in the document]
Comment (fraidev) [reply crates/fluvio-stream-dispatcher/src/metadata/local.rs:602]: Good catch! I forgot to remove it

File: crates/fluvio-stream-dispatcher/src/metadata/local.rs [hunk 3]
```diff
@@ -309,15 +308,17 @@ cfg_if::cfg_if! {
                 Self { path, stores }
             }
 
-            async fn get_store<S: Spec + DeserializeOwned>(&self) -> Result<Arc<SpecStore>> {
+            fn get_store<S: Spec + DeserializeOwned>(&self) -> Result<Arc<SpecStore>> {
```
[the remark this answers is not in the document]
Comment (fraidev) [reply crates/fluvio-stream-dispatcher/src/metadata/local.rs:311]: Maybe not if we plan to move the `SpecStore::load` part to out of this function? 
The `retrieve_items` function still async, that is the most used calling `get_store`.

what do you think?

File: crates/fluvio-stream-dispatcher/src/metadata/local.rs (hunk 1 above)
Comment (fraidev) [reply crates/fluvio-stream-dispatcher/src/metadata/local.rs:319]: https://github.com/infinyon/fluvio/issues/4177

File: crates/fluvio-stream-dispatcher/src/metadata/local.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (fraidev) [reply crates/fluvio-stream-dispatcher/src/metadata/local.rs]: Done, remove it.

File: crates/fluvio-stream-dispatcher/Cargo.toml [hunk 4]
```diff
@@ -26,6 +26,7 @@ once_cell = { workspace = true }
 serde = { workspace = true,  features = ['derive'] }
 serde_json = { workspace = true, optional = true }
 serde_yaml = { workspace = true, optional = true }
+parking_lot = { workspace = true, features = ["send_guard"] }
```
Comment (sehz) [inline crates/fluvio-stream-dispatcher/Cargo.toml:29]: this should be optional

Comment (fraidev) [reply crates/fluvio-stream-dispatcher/Cargo.toml]: Sure! done

**Review by sehz — APPROVED** [review]
LGTM

**Pull request merged**